### ISSUE TITLE
CSI NodePublishVolume readonly is never set

### ISSUE THREAD

**Issue by saad-ali** [kubernetes/kubernetes#61008]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
The `readonly` field passed to `NodePublishVolume` is never set. 

In https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_mounter.go#L198:

```GO
	err = csi.NodePublishVolume(
		ctx,
		c.volumeID,
		c.readOnly,
		deviceMountPath,
		dir,
		accessMode,
		c.volumeInfo,
		attribs,
		nodePublishSecrets,
		fsType,
	)
```

where `c` is a `csiMountMgr` object that is initialized in the `NewMounter(...)` method in https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin.go#L113:

```GO
	mounter := &csiMountMgr{
		plugin:       p,
		k8s:          k8s,
		spec:         spec,
		pod:          pod,
		podUID:       pod.UID,
		driverName:   pvSource.Driver,
		volumeID:     pvSource.VolumeHandle,
		specVolumeID: spec.Name(),
		csiClient:    client,
	}
```

Notice the `readonly` field is never set.

**What you expected to happen**:

The `readonly` field should be set to `Pod.PodSpec.Containers[x].VolumeMounts[x].readonly`

**How to reproduce it (as minimally and precisely as possible)**:
Create a CSI volume and set `Pod.PodSpec.Containers[x].VolumeMounts[x].readonly` to true.
Check the value of `readonly` in the CSI `NodePublishVolume(...)` call.
It should be true, but is false.

CC @vladimirvivien @sbezverk 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**Comment by mlmhl** [comment]
I think `Pod.PodSpec.Containers[x].VolumeMounts[x].Readonly` is only used for mounting pod volume path to container, here(for `volume.Mounter`), we can set `csiMountMgr.readOnly` to `spec.ReadOnly`, just like other volume plugins.

I can help to fix this.

**Comment by sbezverk** [comment]
What about the case when a user wants to mount a volume for one of containers in a POD as read only, but the actual volume has ReadOnly False? I would like to suggest an alternative solution which passes the value of ReadOnly attribute, taken from POD's spec.

**Comment by saad-ali** [comment]
Thank you @mlmhl and @sbezverk for jumping on this!

> I think `Pod.PodSpec.Containers[x].VolumeMounts[x].Readonly` is only used for mounting pod volume path to container, here(for `volume.Mounter`), we can set `csiMountMgr.readOnly` to `spec.ReadOnly`, just like other volume plugins.

I feel like the other volume plugins have the wrong behavior here. The purpose of the `SetupAt` call is to mount the volume. Therefore, the `SetupAt` should receive the pod volume mount readonly field.

That said, as mentioned [here](https://github.com/kubernetes/kubernetes/pull/61014#discussion_r173703179), I think the fact that "`Pod.PodSpec.Containers[x].VolumeMounts[x].Readonly` is used for mounting pod volume path to container" might be hiding a bug: that `spec.Readonly` is never set correctly. I haven't had a chance to validate this but my reading of the code suggests that the `spec.Readonly` is never set. But because the container mount uses `Pod.PodSpec.Containers[x].VolumeMounts[x].Readonly`, even though the plugin did not get the correct readonly value, the mount ends up getting mounted with the correctly readonly value.

Thoughts?

**Comment by saad-ali** [comment]
We discussed the proposed fixes in the Kubernetes CSI implementation meeting today (see [notes](https://docs.google.com/document/d/1-WmRYvqw1FREcD1jmZAOjC0jX6Gop8FMOzsdqXHFoT4/edit#heading=h.l1xczpgfnv2j)). We decided it would be better for CSI to be consistent with the other plugins (and still broken), and have CSI fixed, along with all the other plugins, whenever we fix the `volume.Spec.readonly` issue  (likely post-1.10).

Therefore, I will accept @mlmhl fix and we will close @sbezverk. Then we can remove the 1.10 tag from this issue and track it independent of the 1.10 release.
